jupyter_telemetry/_eventschema.py

```diff
@@ -0,0 +1,110 @@
+from collections import deque
+
+from jsonschema import Draft7Validator, validators
+from jsonschema.exceptions import ValidationError
+
+
+class ExtractCategories(ValidationError):
+    def __init__(self, property, categories, *args, **kwargs):
+        super(ValidationError, self).__init__(*args, **kwargs)
+        self.property = property
+        self.categories = categories
+
+
+def extend_with_categories(validator_class):
+    """
+    Extend the validator class so that during json schema validation, whenever
+    the keyword 'categories' is encountered in a valid context with regards to a
+    property, it yields an instance of ExtractCategories containing the
+    information needed for category filtering later.
+    """
+    validate_properties = validator_class.VALIDATORS["properties"]
+
+    def get_categories(validator, properties, instance, schema):
+        for property, subschema in properties.items():
+            if "categories" in subschema:
+                yield ExtractCategories(property, subschema["categories"], message=None)
+
+        for error in validate_properties(
+            validator, properties, instance, schema,
+        ):
+            yield error
+
+    return validators.extend(
+        validator_class, {"properties": get_categories},
+    )
+
+
+JSONSchemaValidator = Draft7Validator
+CategoryExtractor = extend_with_categories(JSONSchemaValidator)
+
+
+# Ignore categories under any of these jsonschema keywords
+IGNORE_CATEGORIES_SCHEMA_KEYWORDS = {
+    'if', 'not', 'anyOf', 'oneOf', 'then', 'else'
+}
+
+
+def extract_categories_from_errors(errors):
+    for e in errors:
+        if (
+            isinstance(e, ExtractCategories) and
+            not any(p in IGNORE_CATEGORIES_SCHEMA_KEYWORDS
+                    for p in e.absolute_schema_path)
+        ):
+            yield e
+        else:
+            yield from extract_categories_from_errors(e.context)
+
+
+def extract_categories(instance, schema):
+    """
+    Generate dict of ExtractCategories whose keys are pointers to the properties
+    """
+    return {
+        tuple(c.absolute_path + deque([c.property])): c
+        for c in extract_categories_from_errors(
+            CategoryExtractor(schema).iter_errors(instance)
+        )
+    }
+
+
+def filter_categories(instance, categories, allowed_categories, allowed_properties):
```

Comment:
Two comments here:

* I’d prefer we use a more specific function name here. Something like `filter_categories_from_event`. While it’s true that any dictionary instance could presumably work with this function, we’re only promising to support jupyter_telemetry-styled dictionaries here. We should make this function for scoped to our project by making the name specific. Similarly, I’d rename `extract_categories` to `extract_categories_from_event`.
* It probably needs a thorough docstring, especially since categories is a pretty involved data-type. It’s a dictionary of with `CategoryExtractor` as values. `CategoryExtractor`s aren’t exactly clear either when you’re inspecting the API without looking at the codebase directly (i.e. those don’t have helpful docstrings for someone using jupyter_telemetry).

Reply:
Good point. I've renamed them.

My intention was not to expose anything from this categories stuff outside since this implementation is kind of a hack  and we might switch to proper draft-2019/draft-2020 implementation when it's available so I don't want folks to use anything we use internally here.

I've put everything under `_categories.py` and only expose `filter_categories_from_event` via `categories.py` now.

I still added docs as you suggested even for the stuff we don't expose. I agree that they are a bit complicated and having the docs there would help explain what's going on.